From 8e7fc8853a49c594acf2e8395d68b77454570b74 Mon Sep 17 00:00:00 2001
From: Fabrice Gasnier <fabrice.gasnier@foss.st.com>
Date: Fri, 31 Oct 2025 14:48:17 +0100
Subject: [PATCH] phy: stm32: usb2phy: add phy_reset option

On stm32mp21, when exiting pcsi/osi idle state, the PHY may need a reset,
to properly resume operation. This workarounds the missing STPEN bit
in RCC. This is useful, when used with DWC2 controller to initiate a
new gadget session (either start_udc, or drd role update).

Signed-off-by: Fabrice Gasnier <fabrice.gasnier@foss.st.com>
Change-Id: Ifa5cdeced572a8ecfceb5f4abc8b34984f180a19
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/linux-stm32/+/512984
Reviewed-by: Amelie DELAUNAY <amelie.delaunay@foss.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
---
 drivers/phy/st/phy-stm32-usb2phy.c | 17 +++++++++++++++++
 1 file changed, 17 insertions(+)

--- a/drivers/phy/st/phy-stm32-usb2phy.c
+++ b/drivers/phy/st/phy-stm32-usb2phy.c
@@ -533,11 +533,28 @@ static int stm32_usb2phy_phy_power_off(s
 	return 0;
 }
 
+static int stm32_usb2phy_phy_reset(struct phy *phy)
+{
+	struct stm32_usb2phy *phy_dev = phy_get_drvdata(phy);
+
+	/*
+	 * On stm32mp21, when exiting pcsi/osi idle state, the PHY may
+	 * need a reset, to properly resume operation. This allows the
+	 * clock to be suspended in low power state. There's no STPEN
+	 * bit in RCC anyway, to keep the clock active in stop modes.
+	 * This is useful, in conjunction with DWC2 controller to
+	 * initiate a new gadget session (e.g. either udc_start, or
+	 * drd role switch set routine).
+	 */
+	return reset_control_reset(phy_dev->rstc);
+}
+
 static const struct phy_ops stm32_usb2phy_data = {
 	.init = stm32_usb2phy_init,
 	.exit = stm32_usb2phy_exit,
 	.power_on = stm32_usb2phy_phy_power_on,
 	.power_off = stm32_usb2phy_phy_power_off,
+	.reset = stm32_usb2phy_phy_reset,
 	.set_mode = stm32_usb2phy_set_mode,
 	.owner = THIS_MODULE,
 };
